File: vane-velocity/src/main/java/org/oddlama/velocity/listeners/ProxyLoginListener.java
package org.oddlama.velocity.listeners;

import static org.oddlama.velocity.Util.getServerForHost;

import com.google.inject.Inject;
import com.velocitypowered.api.event.Subscribe;
import com.velocitypowered.api.proxy.ProxyServer;
import org.oddlama.vane.proxycore.listeners.LoginEvent;
import org.oddlama.velocity.Velocity;
import org.oddlama.velocity.compat.VelocityCompatServerInfo;
import org.oddlama.velocity.compat.event.VelocityCompatLoginEvent;
import org.oddlama.velocity.compat.event.VelocityCompatPendingConnection;

public class ProxyLoginListener {

    final Velocity velocity;

    @Inject
    public ProxyLoginListener(Velocity velocity) {
        this.velocity = velocity;
    }

    @Subscribe(priority = 0)
    public void login(com.velocitypowered.api.event.connection.LoginEvent event) {
        if (!event.getResult().isAllowed()) return;

        ProxyServer proxy = velocity.getRawProxy();

        final var virtualHost = event.getPlayer().getVirtualHost();
        if (virtualHost.isEmpty()) return;

        final var server = getServerForHost(proxy, virtualHost.get());

        var serverInfo = new VelocityCompatServerInfo(server);
        LoginEvent proxyEvent = new VelocityCompatLoginEvent(
            event,
            velocity,
            serverInfo,
            new VelocityCompatPendingConnection(event.getPlayer())
        );
        proxyEvent.fire();
    }
}

File: vane-velocity/src/main/kotlin/org/oddlama/velocity/listeners/ProxyLoginListener.kt
package org.oddlama.velocity.listeners

import com.google.inject.Inject
import com.velocitypowered.api.event.Subscribe
import com.velocitypowered.api.event.connection.LoginEvent
import org.oddlama.velocity.Util.getServerForHost
import org.oddlama.velocity.Velocity
import org.oddlama.velocity.compat.VelocityCompatServerInfo
import org.oddlama.velocity.compat.event.VelocityCompatLoginEvent
import org.oddlama.velocity.compat.event.VelocityCompatPendingConnection

/**
 * Handles player login events and forwards them to proxy-core login checks.
 *
 * @property velocity plugin instance used for proxy and config access.
 */
class ProxyLoginListener @Inject constructor(private val velocity: Velocity) {
    /**
     * Processes a Velocity login event when the current login result is still allowed.
     *
     * @param event Velocity login event.
     */
    @Subscribe(priority = 0)
    fun login(event: LoginEvent) {
        if (!event.result.isAllowed) return

        val proxy = velocity.rawProxy

        val virtualHost = event.player.virtualHost.orElse(null) ?: return

        val server = getServerForHost(proxy, virtualHost)

        val serverInfo = VelocityCompatServerInfo(server)
        val proxyEvent = VelocityCompatLoginEvent(
            event,
            velocity,
            serverInfo,
            VelocityCompatPendingConnection(event.player)
        )
        proxyEvent.fire()
    }
}
